### PR TITLE
UAH implementation into all existing functions, finalized KDR function (without assists)

### PR DESCRIPTION
**Pull request by bennyrubanov**
Next on my list: 
- need to get the getMateAndAssists function working for identifying the disambiguous mating piece so that i can use that in the KD ratio function to count delivering a checkmate as a kill
- also need to finalize the kdratio function so that it returns the piece with the best kd ratio. might want to break it out into a single game and set of games functions like we did for the distance functions
https://www.notion.so/bennyr/Chess-Analysis-Project-6bc58dcd58da49e7bb2421c630e1aa2c?pvs=4#b2c8f203faf04c7d8e2f307745b18bdf

### CODE REVIEW

File: cjsmin/src/chess.ts [hunk 1]
```diff
@@ -85,7 +69,23 @@ export type UnambiguousPieceSymbol =
...
+  | 'ng'
+  | 'rh'
+  | 'pa'
+  | 'pb'
+  | 'pc'
+  | 'pd'
+  | 'pe'
+  | 'pf'
+  | 'pg'
+  | 'ph';
```
Comment (EllAchE) [inline cjsmin/src/chess.ts:88]: did you just flip the order here?

File: history.json [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  {
+    "color": "b",
+    "captured": "p",
+    "from": "d5",
+    "to": "c3",
+    "flags": "c",
+    "piece": "n",
+    "originalString": "Nxc3#"
+  }
+]
```
Comment (EllAchE) [inline history.json:199]: why this file?

File: tests/metrics.test.ts [hunk 3]
```diff
@@ -284,26 +286,83 @@ describe('getMoveDistanceSingleGame', () => {
       moves: game,
     });
 
-    expect(result.maxDistancePiece).toBe('pe');
+    expect(result.maxDistancePiece).toBe('PE');
```
Comment (EllAchE) [inline tests/metrics.test.ts:289]: which is black and which is white

File: tests/metrics.test.ts [hunk 4]
```diff
@@ -284,26 +286,83 @@ describe('getMoveDistanceSingleGame', () => {
...
+    const game = [
+      {
+        metadata: [],
+        moves: '1. e4 e5 2. d4 exd4 3. Qxd4 Nc6 4. Qa4 Nf6 5. Nc3 d5 6. exd5 Qe7+ 7. Kd1 Bg4+ 8. Kd2 Nxd5 9. Nb5 Ncb4 10. c3 O-O-O 11. f3 Qe3+ 12. Kd1 Nxc3# 0-1'
+      }
+    ];
+
+    const result = await getKillDeathRatios(game);
+
+    console.log('result', result);
```
Comment (EllAchE) [inline tests/metrics.test.ts:306]: rm this

File: tests/metrics.test.ts [hunk 5]
```diff
@@ -284,26 +286,83 @@ describe('getMoveDistanceSingleGame', () => {
...
+
+    expect(result.killsDeathsAssistsMap['pe'].kills).toEqual(1);
+    expect(result.killsDeathsAssistsMap['pe'].deaths).toEqual(1);
```
Comment (EllAchE) [inline tests/metrics.test.ts:309]: can we add the Q & other pieces to this test too? Not a requirement but would be nice. And do you compute it by piece typse?

File: src/metrics/metrics.ts [hunk 6]
```diff
@@ -204,6 +207,7 @@ export async function getMoveDistanceSetOfGames(games: FileReaderGame[]) {
   };
 }
 
+// DONE
```
Comment (EllAchE) [inline src/metrics/metrics.ts:210]: can you rm these?

File: src/metrics/metrics.ts [hunk 7]
```diff
@@ -221,102 +225,131 @@ export function getAverageDistance(
   return { pieceWithHighestAverageDistance, maxAverageDistance };
 }
 
+// IN PROGRESS
```
Comment (EllAchE) [inline src/metrics/metrics.ts:228]: is this? It has a unit test?

File: src/metrics/metrics.ts [hunk 8]
```diff
@@ -221,102 +225,131 @@ export function getAverageDistance(
...
+    const chess = new Chess();
+    const moveGenerator = chess.historyGenerator(game.moves);
+    console.log(`game.moves: ${game.moves}`);
+    const siteLink = game.metadata[1].match(/"(.*?)"/)[1];
+    console.log(`lichess link to game played: ${siteLink}`);
+
+    for (let moveInfo of moveGenerator) {
+      const { move, board } = moveInfo;
+      //console.log("move: ", move)
+      //console.log("board: ", board)
```
Comment (EllAchE) [inline src/metrics/metrics.ts:253]: let's generally clean up all the logs we don't need

File: src/metrics/metrics.ts [hunk 9]
```diff
@@ -221,102 +225,131 @@ export function getAverageDistance(
...
+      lastMove = move;
```
Comment (EllAchE) [inline src/metrics/metrics.ts:254]: I don't understand why you are reassigning move here

File: src/metrics/metrics.ts [hunk 10]
```diff
@@ -221,102 +225,131 @@ export function getAverageDistance(
...
+      console.log(move.originalString)
+      let piece = lastMove.unambiguousSymbol;
+      if (!killsDeathsAssistsMap[piece]) {
+        killsDeathsAssistsMap[piece] = { kills: 0, deaths: 0, assists: 0 };
+      }
```
Comment (EllAchE) [inline src/metrics/metrics.ts:259]: can you jsut initialize this map once?

Comment (EllAchE) [reply src/metrics/metrics.ts:259]: most likely will be faster

**Review by EllAchE — CHANGES_REQUESTED** [review]
I started reviewing but there;s a lot of commented code and stuff. Can you remove that first

File: cjsmin/src/chess.ts (hunk 1 above)
Comment (bennyrubanov) [reply cjsmin/src/chess.ts:88]: yes

File: history.json (hunk 2 above)
Comment (bennyrubanov) [reply history.json:199]: i dont remember adding this in - maybe from one of your PRs?

File: tests/metrics.test.ts (hunk 3 above)
Comment (bennyrubanov) [reply tests/metrics.test.ts:289]: white capitalized, black lower case. realizing that chess.js had it opposite, which is confusing bc it's white capitalized in the chess world

File: src/metrics/metrics.ts (hunk 6 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts:211]: rm what specifically?

File: src/metrics/metrics.ts (hunk 7 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts]: just something used in getMoveDistanceFunct. already in main

File: src/metrics/metrics.ts (hunk 8 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts]: done

File: src/metrics/metrics.ts (hunk 9 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts:254]: need to review later, can remove if unnecessary

File: src/metrics/metrics.ts (hunk 10 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts]: it's initialized twice? are you referring to the statement under the move captured section? if yes, both are necessary bc the initial if (!KDAMap[piece]) doesn't also initialize pieces that are captured before they are moved

File: src/metrics/metrics.ts (hunk 6 above)
Comment (EllAchE) [reply src/metrics/metrics.ts:211]: comments

File: src/metrics/metrics.ts [hunk 11]
```diff
@@ -326,49 +342,107 @@ function checkOpening() {}
...
+
+    if (lastMove.originalString.includes('#')) {
+      matingPiece = lastMove.piece;
+      unambigMatingPiece = lastMove.unambiguousSymbol;
 
-  // This is where we DO need to disambiguate, the same piece type but a different piece could provide the assist.
-  // Pausing further work till we've fixed that
+      // Determine the color of the mated king
+      const matedKingColor = lastMove.color === 'w' ? 'b' : 'w';
+      unambigMatedPiece = matedKingColor === 'w' ? 'K' : 'k';
```
Comment (EllAchE) [inline src/metrics/metrics.ts:389]: I am confused why we need all of these variables, maybe we only need the last one?

File: src/metrics/metrics.ts [hunk 12]
```diff
@@ -326,49 +342,107 @@ function checkOpening() {}
...
+export function gameHistoryToPgn(gameHistory: GameHistoryMove[]): string {
+  const chess = new Chess();
+
+  for (const move of gameHistory) {
+    chess.move(move);
+  }
+
+  return chess.pgn();
+}
+
```
Comment (EllAchE) [inline src/metrics/metrics.ts:445]: These are helpers for the tests right? Can we put them there?

File: src/metrics/metrics.ts [hunk 13]
```diff
@@ -65,79 +66,81 @@ export function countGamesInDataset(datasetPath: string): number {
...
+        if (moveInfo.move.flags === 'q') {
+          distanceMap['K'] += 2
+          distanceMap['RA'] = (distanceMap['RA'] || 0) + 3
+        }
+        // kingside castle
+        else {
+          distanceMap['K'] += 2
+          distanceMap['RH'] = (distanceMap['RH'] || 0) + 2
+        }
+      }
```
Comment (EllAchE) [inline src/metrics/metrics.ts:127]: Not strictly necessary but this if is hard to follow. I'd propose this to make it more readable?
```suggestion
let movingKing =  'k'
let movingRook, rookDistance;

if (moveInfo.move.flags === 'k') {
rookDistance = 2
movingRook = 'ra'
} else {
rookDistance = 3
movingRook = 'rh'
}
if (moveInfo.move.color === 'w') {
movingRook = movingRook.toUpperCase()
movingKing = movingKing.toUpperCase()
}

distanceMap[movingKing] += 2
distanceMap[movingRook] += rookDistance
```
Something along these lines

Comment (EllAchE) [reply src/metrics/metrics.ts:127]: hacked this together, this is highely opinionated so dw too much but having less lines and no nesting makes it easier to follow imo

File: src/metrics/metrics.ts [hunk 14]
```diff
@@ -65,79 +66,81 @@ export function countGamesInDataset(datasetPath: string): number {
...
+    let piece = lastMove.unambiguousSymbol;
+
+    if (!distanceMap[piece]) {
+      distanceMap[piece] = 0;
+    }
+
+    const fromMove = moveInfo.move.from
+    const toMove = moveInfo.move.to
+
+    let distance = 0;
```
Comment (EllAchE) [inline src/metrics/metrics.ts:98]: pretty confusing bc this isn't used until the else block. Let's move it there and take out the redundant reassigning

File: src/metrics/metrics.ts [hunk 15]
```diff
@@ -221,102 +226,113 @@ export function getAverageDistance(
...
-    for (const move of moveHistory) {
-      if (move.captured) {
-        if (!kills[move.piece]) {
-          kills[move.piece] = 0;
-        }
-        kills[move.piece]++;
+    const chess = new Chess();
+    const moveGenerator = chess.historyGenerator(game.moves);
+    const siteLink = game.metadata[1].match(/"(.*?)"/)[1];
+    console.log(`lichess link to game played: ${siteLink}`);
```
Comment (EllAchE) [inline src/metrics/metrics.ts:249]: useful to have this info fs

File: src/metrics/metrics.ts [hunk 16]
```diff
@@ -326,49 +342,107 @@ function checkOpening() {}
...
+    const { move, board } = moveInfo;
+
+    lastMove = move;
+
+    // Add the current move to the start of the array
+    lastFewMoves.unshift(move);
+
+    // If we have more than 5 moves in the array, remove the oldest one
+    if (lastFewMoves.length > 5) {
+      lastFewMoves.pop();
```
Comment (EllAchE) [inline src/metrics/metrics.ts:374]: this is nbd rn but for performance we can just track total number of moves and add from there, or run reversed hisstory

File: tests/metrics.test.ts (hunk 3 above)
Comment (EllAchE) [reply tests/metrics.test.ts:298]: so which are doing, now i am confused

Comment (bennyrubanov) [reply tests/metrics.test.ts:298]: right now code for unambig pieces between chess.ts and what we have written is structured for white to be capitalized, no need to change anything unless prefer otherwise

getMateAndAssists function also working properly for ambig pieces, haven't tested other functions or other functions aren't using ambig pieces

File: tests/metrics.test.ts (hunk 5 above)
Comment (bennyrubanov) [reply tests/metrics.test.ts:318]: can. computed by piece unambig piece type rn yes

File: src/metrics/metrics.ts (hunk 6 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts:211]: why?

File: src/metrics/metrics.ts (hunk 11 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts:389]: unambigMatingPiece used in KDR function to update kill count for the mating piece
matingPiece defined just to have if we want ambig mating piece kdrs down the line
matedKingColor just used to identify which king was mated, which is then used in KDR funct to update death count for the mated king

File: src/metrics/metrics.ts (hunk 12 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts:445]: yes. i think first used them in metrics.ts for a different function but i dont think they are used there anymore. but any reason to move them over?

File: src/metrics/metrics.ts (hunk 13 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts:127]: yeah fair enough. smart solution for sure but idk how much i care to do stuff like this, feel free if you have high preference. i do think urs is more elegant

File: src/metrics/metrics.ts (hunk 14 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts:98]: what specifically? all these lines?

File: src/metrics/metrics.ts (hunk 16 above)
Comment (bennyrubanov) [reply src/metrics/metrics.ts:374]: true. not going to focus on overengineering for now but can refactor later

**Review by EllAchE — APPROVED** [review]
